fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+def connect_to_gcs():
+    subprocess.run(['gcloud', 'auth', 'activate-service-account', '--key-file', GCS_CREDENTIALS_FILE], check=True)
+    return google.cloud.storage.Client.from_service_account_json(GCS_CREDENTIALS_FILE)
+    #return google.cloud.storage.Client(project = 'near-nayduck')
+
+NUM_FUZZERS = os.cpu_count()
+#NUM_FUZZERS = 1
+
+CMD_PORT = 7055
+METRICS_PORT = 5507
+
+REPO_DIR_PARENT = WORKDIR
+REPO_DIR_NAME = 'fuzzed-nearcore'
+REPO_DIR = REPO_DIR_PARENT / REPO_DIR_NAME
+
+GCS_CREDENTIALS_FILE = config.CONFIG_DIR / 'credentials.json'
+GCS_BUCKET = 'fuzzer'
+
+FUZZ_TIME = prometheus_client.Counter('fuzz_seconds', 'Time spent fuzzing', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CRASHES = prometheus_client.Counter('fuzz_crashes', 'Number of times the fuzzer process crashed (not unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_ARTIFACTS_FOUND = prometheus_client.Counter('fuzz_artifacts_found', 'Number of artifacts found (should be number of unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_UPLOADED = prometheus_client.Counter('fuzz_corpus_uploaded', 'Number of elements uploaded to GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_DELETED = prometheus_client.Counter('fuzz_corpus_deleted', 'Number of elements deleted from GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+
+def update_repo(branch):
+    if not REPO_DIR.exists():
+        print(f"Doing initial clone of repository {REPO_DIR}")
+        if 0 != subprocess.run(['git', 'clone', REPO_URL, REPO_DIR_NAME], cwd=REPO_DIR_PARENT).returncode:
+            raise RuntimeError(f"Failed to clone from repo {REPO_URL}")
+
+    print(f"Updating to latest commit of branch {branch}")
+    if 0 != subprocess.run(['git', 'checkout', '-f', branch], cwd=REPO_DIR).returncode:
+        raise RuntimeError(f"Failed to checkout branch {branch}")
+    if 0 != subprocess.run(['git', 'pull', '--ff-only', REPO_URL, branch], cwd=REPO_DIR).returncode:
+        raise RuntimeError(f"Failed to fast-forward branch {branch} to latest from repo {REPO_URL}")
+
+def parse_config():
+    return toml.load(REPO_DIR / 'fuzz.toml')
+
+FUZZERS = []
+
+def report_artifact(gcs_path, branch, t):
+    # todo: report on zulip with all the details
+    FUZZ_ARTIFACTS_FOUND.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc()
+
+def artifact_deleted():
+    raise RuntimeError(f"Deleted artifact {path}, which should have been prevented by ignore_deletions")
+
+def run_fuzzer(branch, t, gcs, exit_event):
+    # First, figure out the current corpus version
+    bucket = gcs.get_bucket(GCS_BUCKET)
+    corpus_vers = bucket.blob('current-corpus').download_as_text().strip()
+
+    log_path = (datetime.datetime.now().strftime('%Y-%m-%d') + '/' +
+        t['crate'] + '/' +
+        t['runner'] + '/' +
+        str(uuid.uuid4()))
+    log_filename = WORKDIR / 'fuzz-logs' / log_path
+    os.makedirs(log_filename.parent, exist_ok=True)
```

Comment:
It felt to me like it was something mostly for when one needs to create multiple directories, as for a single directory it's just an additional layer of indirection when trying to understand the code. Would you like it better if `utils.mkdirs` were used in the 3 places where a single makedirs is currently used here?

Reply:
I don’t have strong feelings. It does hide the `exist_ok` so that’s why I kinda prefer it. At the moment the function is only ever used to create a single directory.

Reply:
Got it, switching to that :)